src/order/well_founded_set.lean

```diff
@@ -147,11 +154,171 @@ end
+  rw [subtype.val_eq_coe, h] at hx,
+  apply add_right_cancel (hx.trans y.2.1.symm),
+end⟩
+
+lemma eq_of_fst_eq_fst {x y : (add_antidiagonal s t a)}
+  (h : (x : α × α).fst = (y : α × α).fst) : x = y :=
+subtype.ext (prod.ext h (add_antidiagonal.fst_eq_fst_iff_snd_eq_snd.1 h))
+
+lemma eq_of_snd_eq_snd {x y : (add_antidiagonal s t a)}
+  (h : (x : α × α).snd = (y : α × α).snd) : x = y :=
+subtype.ext (prod.ext (add_antidiagonal.fst_eq_fst_iff_snd_eq_snd.2 h) h)
+
+end add_antidiagonal
 end set
+
+namespace set
+namespace add_antidiagonal
+
+variables [linear_ordered_cancel_add_comm_monoid α] (s t : set α) (a : α)
+
+/-- The relation on `set.add_antidagonal s t a` given by `<` on the first coordinate. -/
+def lt_left (x y : add_antidiagonal s t a) : Prop := (x : α × α).1 < (y : α × α).1
+
+/-- `set.add_antidiagonal s t a` ordered by `lt_left` embeds into `s` -/
+def fst_rel_embedding : (lt_left s t a) ↪r ((<) : s → s → Prop) :=
+⟨⟨λ x, ⟨(↑x : α × α).1, x.2.2.1⟩, λ x y hxy,
+    add_antidiagonal.eq_of_fst_eq_fst (subtype.mk_eq_mk.1 hxy)⟩, λ x y, iff.refl _⟩
+
+/-- `set.add_antidiagonal s t a` ordered by `lt_left` embeds into the dual of `t` -/
+def snd_rel_embedding : (lt_left s t a) ↪r ((>) : t → t → Prop) :=
+⟨⟨λ x, ⟨(↑x : α × α).2, x.2.2.2⟩, λ x y hxy,
+    add_antidiagonal.eq_of_snd_eq_snd (subtype.mk_eq_mk.1 hxy)⟩, λ x y, begin
+  simp only [lt_left, subtype.mk_lt_mk, gt_iff_lt, function.embedding.coe_fn_mk],
+  split; intro h,
+  { by_contra hle,
+    rw not_lt at hle,
+    have h := add_lt_add_of_le_of_lt hle h,
+    rw [← subtype.val_eq_coe, ← subtype.val_eq_coe, x.2.1, y.2.1] at h,
+    exact lt_irrefl a h },
+  { by_contra hle,
+    rw not_lt at hle,
+    have h := add_lt_add_of_le_of_lt hle h,
+    rw [← subtype.val_eq_coe, ← subtype.val_eq_coe, add_comm, x.2.1, add_comm, y.2.1] at h,
+    exact lt_irrefl a h }
+end⟩
+
+variables {s} {t}
+
+theorem finite_of_is_wf (hs : s.is_wf) (ht : t.is_wf) (a) :
+  (add_antidiagonal s t a).finite :=
+begin
+  by_contra h,
+  rw [← set.infinite, ← infinite_coe_iff] at h,
+  haveI := h,
+  haveI : is_well_order s (<) := {
+    wf := hs,
+    .. (infer_instance : is_strict_total_order _ _),
+  },
+  haveI : is_well_order (add_antidiagonal s t a) (lt_left s t a) :=
+    (add_antidiagonal.fst_rel_embedding s t a).is_well_order,
+  have hwf : well_founded
+    (function.swap (lt_left s t a)) :=
+    (add_antidiagonal.snd_rel_embedding s t a).swap.well_founded ht,
+  exact not_well_founded_swap_of_infinite_of_well_order (hwf),
+end
+
+end add_antidiagonal
+end set
+
+namespace finset
+
+variables [linear_ordered_cancel_add_comm_monoid α]
+variables {s t : set α} (hs : s.is_wf) (ht : t.is_wf) (a : α)
+
+/-- `finset.add_antidiagonal_of_is_wf hs ht a` is the set of all pairs of an element in
+  `s` and an element in `t` that add to `a`, but its construction requires proofs `hs` and `ht` that
+  `s` and `t` are well-ordered. -/
+noncomputable def add_antidiagonal_of_is_wf : finset (α × α) :=
```

Comment:
I think I would maybe even leave out the `_of_is_wf` from the name. Short names for definitions are good, because long names imply long names for all lemmas about that concept.